lokimq/bt_serialize.h

```diff
@@ -909,5 +922,192 @@ class bt_dict_consumer : private bt_list_consumer {
+        char* op = begin;
+
+        *(begin++) = (size % 10) + '0';
+        size /= 10;
+
+        while (size > 0)
+        {
+            *(begin++) = (size % 10) + '0';
+            size /= 10;
+            digits++;
+        }
+
+        char* ed = begin - 1;
+        while (ed > op)
+        {
+            char tmp = *ed;
+            *ed-- = *op;
+            *op++ = tmp;
+        }
+
+        return digits;
+    }
+
+    template <typename T>
+    fixed_buffer_producer& append_integer(T number)
+    {
+        static_assert(std::is_integral<T>::value, "Can't BEncode non-integral type as integer.");
+
+        // integer takes min space 3, e.g. i3e or i0e
+        check_space(3);
+        *(begin++) = 'i';
+
+        char* op = begin;
+
+        // first digit always present, even if value is 0
+        *(begin++) = (number % 10) + '0';
+        number /= 10;
+
+        // will have to profile, but one assumes it is faster to check per-digit than
+        // to calculate number of digits then check.
+        while (number > 0)
+        {
+            check_space(2); // need space for the 'e' to end the number, as well as current digit
+            *(begin++) = (number % 10) + '0';
+            number /= 10;
+        }
+
+        char *ed = begin - 1;
+        while (ed > op)
+        {
+            char tmp = *ed;
+            *ed-- = *op;
+            *op++ = tmp;
+        }
+
+        *(begin++) = 'e';
+
+        return *this;
+    }
+
+    template <typename T>
+    fixed_buffer_producer& append_integer(T number, std::string_view key)
+    {
+        static_assert(std::is_integral<T>::value, "Can't BEncode non-integral type as integer.");
+
+        append_string(key);
+        append_integer(number);
+        return *this;
+    }
+
+    fixed_buffer_producer& append_string(std::string_view string)
+    {
+        size_t digits = num_digits(string.size());
+
+        check_space(digits + string.size() + 1);
+
+        append_size(string.size());
+
+        *(begin++) = ':';
+
+        memcpy(begin, string.data(), string.size());
+        begin += string.size();
+        return *this;
+    }
+
+    fixed_buffer_producer& append_string(std::string_view string, std::string_view key)
+    {
+        append_string(key);
+        append_string(string);
+        return *this;
+    }
+
+    fixed_buffer_producer& append_serialized_data(std::string_view string)
+    {
+        check_space(string.size());
+        memcpy(begin, string.data(), string.size());
+        begin += string.size();
+        return *this;
+    }
+
+    fixed_buffer_producer& start_list()
+    {
+        check_space(2);
+        *(begin++) = 'l';
+        end--;
+        return *this;
+    }
+
+    fixed_buffer_producer& start_list(std::string_view key)
+    {
+        append_string(key);
+        return start_list();
+    }
+
+    fixed_buffer_producer& start_dict()
+    {
+        check_space(2);
+        *(begin++) = 'd';
+        end--;
+        return *this;
+    }
```

Comment:
I'm not a fan of the way this works because the caller will have a `fixed_buffer_producer`, but will *also* have a bunch of state associated with it that they have to keep track of -- is it making a dict or a list? Have I written a key yet? Do I have a sublist/subdict active?

One idea to make this better is to have RAII subclasses for nested types:

```C++
// Existing classes reworked a bit:
class fixed_buffer_base {
protected:
    ...

    int nesting_ = 0;
    void end_nesting(char* new_begin) { begin = new_begin; }
    virtual void begin_nesting() { nesting_++; }
    friend class nested_dict;
    friend class nested_list;
};
class fixed_dict_buffer : public fixed_buffer_base {
    ...
    nested_dict begin_dict(std::string_view key) {
        // check nesting_ == 0
        // check minimum size available for key + empty dict
        // check key order
        // write key
        // update last_key
        // (all of the above should just be some method since they'll be identical checks for every kv pair)

        return {*this};
    }
};
class fixed_list_buffer : public fixed_buffer_base { ... };

// New nested classes:
class nested_dict : fixed_dict_buffer {
    fixed_buffer_base& b;
    nested_dict(fixed_buffer_base& b) : b{b} {
        // Copy buffer pointers from b
        b.begin_nesting();
        b.begin_dict();
    }
    ~nested_dict() {
        b.end_dict();
        b.end_nesting(begin);
    }
    void begin_nesting() override { nesting_++; b.begin_nesting(); }
};
class nested_list : fixed_list_buffer {
    fixed_buffer_base& b;
    nested_dict(fixed_buffer_base& b) : b{b} {
        b.nesting_++;
        b.begin_list();
    }
    ~nested_list() {
        b.end_list();
        b.end_nesting(begin);
    }
    void begin_nesting() override { nesting_++; b.begin_nesting(); }
};
```

Now the caller does something like:

```C++

fixed_dict_buffer foo{start, end};
foo.append_string("foo", "bar");
{
    auto x = foo.append_dict("b");
    // foo.append_string("xyz", "abc");  -- would throw because there is an open nested structure
    x.append_string("c", "xyz");
    {
        auto y = x.append_list("d");
        y.append_integer(42);
     }
}
foo.append_string("xyz", "abc");

Reply:
a potential bug in that example code: the nested_dict and nested_list destructors may throw

Reply:
True; it should probably fiddle with the end pointer (i.e. shorten it by one) so that there is always space to write the 'e' during destruction.